libraries/AP_RangeFinder/AP_RangeFinder_SRF02I2C.cpp

```diff
@@ -0,0 +1,162 @@
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include "AP_RangeFinder_SRF02I2C.h"
+
+#include <utility>
+
+#include <AP_HAL/AP_HAL.h>
+#include <AP_HAL/utility/sparse-endian.h>
+
+extern const AP_HAL::HAL& hal;
+
+/**
+ * Constructor
+ *
+ */
+AP_RangeFinder_SRF02I2C::AP_RangeFinder_SRF02I2C(RangeFinder::RangeFinder_State &_state, AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev)
+            : AP_RangeFinder_Backend(_state)
+            , _dev(std::move(dev)){}
+
+/**
+ * Detect
+ *
+ * @return nullptr is Sensor not found or Initialization failure. <BR>
+ * Other is Instance.
+ */
+AP_RangeFinder_Backend *AP_RangeFinder_SRF02I2C::detect(RangeFinder::RangeFinder_State &_state, AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev)
+{
+    AP_RangeFinder_SRF02I2C *sensor = new AP_RangeFinder_SRF02I2C(_state, std::move(dev));
+
+    // Device initialization
+    if (!sensor || !sensor->_init()) {
+        delete sensor;
+        return nullptr;
+    }
+
+    return sensor;
+}
+
+/**
+ * update the state of the sensor
+ *
+ */
+void AP_RangeFinder_SRF02I2C::update(void)
+{
+    if (_sem->take_nonblocking()) {
+        if (_new_distance) {
+            // Measurement valid
+            state.distance_cm = _distance;
+            _new_distance = false;  // Set invalid measurement
+            update_status();
+        } else {
+            // Invalid measurement
+            set_status(RangeFinder::RangeFinder_NoData);
+        }
+        _sem->give();
+    }
+}
+
+/**
+ * Initialization
+ *
+ * @retval true Successful completion
+ * @retval false Abnormal termination
+ */
+bool AP_RangeFinder_SRF02I2C::_init()
+{
+    if (!_dev->get_semaphore()->take(HAL_SEMAPHORE_BLOCK_FOREVER)) {
+        return false;
+    }
+
+    // Issuing measurement start command
+    bool ret = _startMeasurement();
+
+    _dev->get_semaphore()->give();
+
+    if (ret) {
+        // call timer() at 14Hz
+        _dev->register_periodic_callback(70000,
+                                         FUNCTOR_BIND_MEMBER(&AP_RangeFinder_SRF02I2C::_timer, void));
+    }
+
+    return ret;
+}
+
+/**
+ * Instruct start of measurement
+ *
+ * @retval true Successful measurement start instruction
+ * @retval false Measurement start instruction failure (with retry)
+ */
+bool AP_RangeFinder_SRF02I2C::_startMeasurement(void)
+{
+    // Set Real range mode and Set unit cm
+    static const uint8_t cmd[2] = {SRF02_I2C_REG_W_COMMAND, SRF02_I2C_CMD_REAL_RANGE_MODE_RESULT_IN_CENTIMETERS};
+
+    _dev->set_retries(10);
+    return _dev->transfer(cmd, sizeof(cmd), nullptr, 0);
```

Comment:
I know that. I have not confirmed the specifications.
I confirmed the specification.
In case of 1 register writing and register reading continuously, change to write_register, read_registers.